File: tests/test_parser.c
#include "acutest.h"
#include "../src/core/internal.h"
#include <string.h>

// Declaración de la función a testear
void parse_line(const char *raw, SlideLine *out);

void test_parse_title(void) {
    SlideLine sl;
    parse_line("# Mi Titulo", &sl);
    TEST_CHECK(sl.type == LINE_TITLE);
    TEST_CHECK(strcmp(sl.text, "Mi Titulo") == 0);
}

void test_parse_subtitle(void) {
    SlideLine sl;
    parse_line("## Mi Subtitulo", &sl);
    TEST_CHECK(sl.type == LINE_SUBTITLE);
    TEST_CHECK(strcmp(sl.text, "Mi Subtitulo") == 0);
}

void test_parse_blockquote(void) {
    SlideLine sl;
    parse_line("> Nota importante", &sl);
    TEST_CHECK(sl.type == LINE_BLOCKQUOTE);
    TEST_CHECK(strcmp(sl.text, "Nota importante") == 0);
}

void test_parse_bullets(void) {
    SlideLine sl;
    // Nivel 1
    parse_line("- Item 1", &sl);
    TEST_CHECK(sl.type == LINE_BULLET1);
    TEST_CHECK(strcmp(sl.text, "Item 1") == 0);

    // Nivel 2
    parse_line("  - Item 2", &sl);
    TEST_CHECK(sl.type == LINE_BULLET2);
    TEST_CHECK(strcmp(sl.text, "Item 2") == 0);
}

void test_parse_image(void) {
    SlideLine sl;
    // Formato corto
    parse_line("!logo.png", &sl);
    TEST_CHECK(sl.type == LINE_IMAGE);
    TEST_CHECK(strcmp(sl.text, "logo.png") == 0);

    // Formato Markdown completo
    parse_line("![alt text](path/to/img.png)", &sl);
    TEST_CHECK(sl.type == LINE_IMAGE);
    TEST_CHECK(strcmp(sl.text, "path/to/img.png") == 0);
}

void test_parse_table(void) {
    SlideLine sl;
    // Fila normal
    parse_line("| A | B |", &sl);
    TEST_CHECK(sl.type == LINE_TABLE_ROW);
    TEST_CHECK(sl.ncols == 2);
    TEST_CHECK(strcmp(sl.cols[0], "A") == 0);
    TEST_CHECK(strcmp(sl.cols[1], "B") == 0);

    // Separador
    parse_line("|---|---|", &sl);
    TEST_CHECK(sl.type == LINE_TABLE_SEP);
}

void test_parse_task_list(void) {
    SlideLine sl;
    // Unchecked
    parse_line("- [ ] Tarea pendiente", &sl);
    TEST_CHECK(sl.type == LINE_TASK_UNCHECKED);
    TEST_CHECK(strcmp(sl.text, "Tarea pendiente") == 0);

    // Checked
    parse_line("- [x] Tarea completada", &sl);
    TEST_CHECK(sl.type == LINE_TASK_CHECKED);
    TEST_CHECK(strcmp(sl.text, "Tarea completada") == 0);

    // Checked uppercase
    parse_line("- [X] Tarea completada 2", &sl);
    TEST_CHECK(sl.type == LINE_TASK_CHECKED);
    TEST_CHECK(strcmp(sl.text, "Tarea completada 2") == 0);
}

TEST_LIST = {
    { "parse_title", test_parse_title },
    { "parse_subtitle", test_parse_subtitle },
    { "parse_blockquote", test_parse_blockquote },
    { "parse_bullets", test_parse_bullets },
    { "parse_image", test_parse_image },
    { "parse_table", test_parse_table },
    { "parse_task_list", test_parse_task_list },
    { NULL, NULL }
};

File: python/test_parser.py
import unittest
import os
from c_slides_bindings import CSlides, LineType

class TestParser(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        lib_path = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "libslider.so"))
        cls.cs = CSlides(lib_path)

    def test_parse_title(self):
        sl = self.cs.parse_line("# Mi Titulo")
        self.assertEqual(sl.type, LineType.TITLE)
        self.assertEqual(sl.text.decode('utf-8'), "Mi Titulo")

    def test_parse_subtitle(self):
        sl = self.cs.parse_line("## Mi Subtitulo")
        self.assertEqual(sl.type, LineType.SUBTITLE)
        self.assertEqual(sl.text.decode('utf-8'), "Mi Subtitulo")

    def test_parse_blockquote(self):
        sl = self.cs.parse_line("> Nota importante")
        self.assertEqual(sl.type, LineType.BLOCKQUOTE)
        self.assertEqual(sl.text.decode('utf-8'), "Nota importante")

    def test_parse_bullets(self):
        # Nivel 1
        sl = self.cs.parse_line("- Item 1")
        self.assertEqual(sl.type, LineType.BULLET1)
        self.assertEqual(sl.text.decode('utf-8'), "Item 1")

        # Nivel 2
        sl = self.cs.parse_line("  - Item 2")
        self.assertEqual(sl.type, LineType.BULLET2)
        self.assertEqual(sl.text.decode('utf-8'), "Item 2")

    def test_parse_image(self):
        # Formato corto
        sl = self.cs.parse_line("!logo.png")
        self.assertEqual(sl.type, LineType.IMAGE)
        self.assertEqual(sl.text.decode('utf-8'), "logo.png")

        # Formato Markdown completo
        sl = self.cs.parse_line("![alt text](path/to/img.png)")
        self.assertEqual(sl.type, LineType.IMAGE)
        self.assertEqual(sl.text.decode('utf-8'), "path/to/img.png")

    def test_parse_table(self):
        # Fila normal
        sl = self.cs.parse_line("| A | B |")
        self.assertEqual(sl.type, LineType.TABLE_ROW)
        self.assertEqual(sl.ncols, 2)
        self.assertEqual(sl.cols[0].value.decode('utf-8'), "A")
        self.assertEqual(sl.cols[1].value.decode('utf-8'), "B")

        # Separador
        sl = self.cs.parse_line("|---|---|")
        self.assertEqual(sl.type, LineType.TABLE_SEP)

    def test_parse_task_list(self):
        # Unchecked
        sl = self.cs.parse_line("- [ ] Tarea pendiente")
        self.assertEqual(sl.type, LineType.TASK_UNCHECKED)
        self.assertEqual(sl.text.decode('utf-8'), "Tarea pendiente")

        # Checked
        sl = self.cs.parse_line("- [x] Tarea completada")
        self.assertEqual(sl.type, LineType.TASK_CHECKED)
        self.assertEqual(sl.text.decode('utf-8'), "Tarea completada")

        # Checked uppercase
        sl = self.cs.parse_line("- [X] Tarea completada 2")
        self.assertEqual(sl.type, LineType.TASK_CHECKED)
        self.assertEqual(sl.text.decode('utf-8'), "Tarea completada 2")

if __name__ == "__main__":
    unittest.main()
